enhancements/20220214-harvester-cloud-provider-enhancement.md

```diff
@@ -0,0 +1,237 @@
+  name: vlan100-pool
+spec:
+  selector:
+    priority: 100
+    scope:
+    - project: ""
+      namespace: default
+      guestCluster: '*'
+    - project: ""
+      namespace: cattle-fleet-system
+      guestCluster: '*'
+  ranges:
+  - subnet: 192.168.41.0/24
+status:
+  allocated:
+    192.168.41.4: default/kubernetes-default-nginx-lb-3-3be238b0
+    192.168.41.5: default/kubernetes-default-lb20-e99740b0
+  allocatedHistory:
+    192.168.41.2: default/kubernetes-default-nginx-lb-cwa-af13e48f
+    192.168.41.3: default/kubernetes-default-nginx-lb-3-9d7d8a7a
+  available: 251
+  conditions:
+  - lastUpdateTime: "2022-05-20T07:41:36Z"
+    status: "True"
+    type: Ready
+  lastAllocated: 192.168.41.4
+  total: 253
+```
+
+Modify LoadBalancer CRD
+```yaml
+apiVersion: loadbalancer.harvesterhci.io/v1alpha1
+kind: LoadBalancer
+metadata:
+  name: lb1
+  namespace: default
+spec:
+  backendServerSelector:
+    demo: demo
+  healthCheck: {}
+  ipPool: vlan-awareness-pool
+  ipam: pool
+  listeners:
+  - backendPort: 80
+    name: ""
+    port: 80
+    protocol: TCP
+  workloadType: vm
+status:
+  address: 172.16.231.10
+  allocatedAddress:
+    gateway: 172.16.231.1
+    ip: 172.16.231.10
+    ipPool: vlan-awareness-pool
+    mask: 255.255.255.0
+  conditions:
+  - lastUpdateTime: "2022-05-20T07:55:20Z"
+    status: "True"
+    type: Ready
+```
+
+```yaml
+apiVersion: loadbalancer.harvesterhci.io/v1alpha1
+kind: LoadBalancer
+metadata:
+  annotations:
+    loadbalancer.harvesterhci.io/cluster: kubernetes
+    loadbalancer.harvesterhci.io/namespace: default
+    loadbalancer.harvesterhci.io/network: default/vlan100
+    loadbalancer.harvesterhci.io/project: local:p-hl965
+  labels:
+    cloudprovider.harvesterhci.io/cluster: kubernetes
+    cloudprovider.harvesterhci.io/serviceName: nginx-lb-3
+    cloudprovider.harvesterhci.io/serviceNamespace: default
+  name: kubernetes-default-nginx-lb-3-3be238b0
+  namespace: default
+spec:
+  ipam: pool
+  workloadType: cluster
+status:
+  address: 192.168.41.4
+  allocatedAddress:
+    gateway: 192.168.41.1
+    ip: 192.168.41.4
+    ipPool: vlan100-pool
+    mask: 255.255.255.0
+  conditions:
+  - lastUpdateTime: "2022-05-21T14:21:53Z"
+    status: "True"
+    type: Ready
+```
+
+## Design
+
+### Implementation Overview
+
+#### load balancer types
+
+There are two types of load balancers offered by Harvester cloud providers based on their usage.
+
+- LB for application deployed in the guest cluster(cluster LB): The cloud provider creates the LB when watching a service in the guest cluster. The load balancer IP is exposed in the VLAN of the guest cluster where we deploy the kube-vip.
+- LB for VM workloads(VM LB): It is for traditional VM workloads deployed on Harvester VMs. The load balancer IP is exposed in the LAN of the Harvester. The Harvester will forward the traffic to a Kubernetes service whose endpoints are the nodeport address of the load balancer service in the guest cluster.
```

Comment:
could those VM workloads also be running in specific VLANs ?

Reply:
No, it's not supported yet.